File: lib/flutter/test/didkit_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:didkit/didkit.dart';
import 'dart:convert';
import 'package:uuid/uuid.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  test('getVersion', () async {
    expect(DIDKit.getVersion(), isInstanceOf<String>());
  });

  test('exceptions', () async {
    expect(() => DIDKit.issueCredential('', '', ''), throwsA(isInstanceOf<DIDKitException>()));
    expect(() => DIDKit.issuePresentation('', '', ''), throwsA(isInstanceOf<DIDKitException>()));
    expect(() => DIDKit.verifyCredential('', ''), throwsA(isInstanceOf<DIDKitException>()));
    expect(() => DIDKit.verifyPresentation('', ''), throwsA(isInstanceOf<DIDKitException>()));
  });

  test('generateEd25519Key', () async {
    expect(DIDKit.generateEd25519Key(), isInstanceOf<String>());
  });

  test('keyToDID', () async {
    final key = DIDKit.generateEd25519Key();
    final did = DIDKit.keyToDID('key', key);
    expect(did, isInstanceOf<String>());
  });

  test('verificationMethod', () async {
    final key = DIDKit.generateEd25519Key();
    final vm = DIDKit.keyToVerificationMethod('key', key);
    expect(vm, isInstanceOf<String>());
  });

  test('issueCredential, verifyCredential', () async {
    final key = DIDKit.generateEd25519Key();
    final did = DIDKit.keyToDID('key', key);
    final verificationMethod = DIDKit.keyToVerificationMethod('key', key);
    final options = {
        'proofPurpose': 'assertionMethod',
        'verificationMethod': verificationMethod
    };
    final credential = {
        '@context': 'https://www.w3.org/2018/credentials/v1',
        'id': 'http://example.org/credentials/3731',
        'type': ['VerifiableCredential'],
        'issuer': did,
        'issuanceDate': '2020-08-19T21:41:50Z',
        'credentialSubject': {
           'id': 'did:example:d23dd687a7dc6787646f2eb98d0'
        }
    };
    final vc = DIDKit.issueCredential(jsonEncode(credential), jsonEncode(options), key);

    final verifyOptions = {
        'proofPurpose': 'assertionMethod'
    };
    final verifyResult = jsonDecode(DIDKit.verifyCredential(vc, jsonEncode(verifyOptions)));
    expect(verifyResult['errors'], isEmpty);
  });

  test('issuePresentation, verifyPresentation', () async {
    final key = DIDKit.generateEd25519Key();
    final did = DIDKit.keyToDID('key', key);
    final verificationMethod = DIDKit.keyToVerificationMethod('key', key);
    final options = {
        'proofPurpose': 'authentication',
        'verificationMethod': verificationMethod
    };
    final presentation = {
        '@context': ['https://www.w3.org/2018/credentials/v1'],
        'id': 'http://example.org/presentations/3731',
        'type': ['VerifiablePresentation'],
        'holder': did,
        'verifiableCredential': {
            '@context': 'https://www.w3.org/2018/credentials/v1',
            'id': 'http://example.org/credentials/3731',
            'type': ['VerifiableCredential'],
            'issuer': 'did:example:30e07a529f32d234f6181736bd3',
            'issuanceDate': '2020-08-19T21:41:50Z',
            'credentialSubject': {
                'id': 'did:example:d23dd687a7dc6787646f2eb98d0'
            }
        }
    };
    final vc = DIDKit.issuePresentation(jsonEncode(presentation), jsonEncode(options), key);

    final verifyOptions = {
        'proofPurpose': 'authentication'
    };
    final verifyResult = jsonDecode(DIDKit.verifyPresentation(vc, jsonEncode(verifyOptions)));
    expect(verifyResult['errors'], isEmpty);
  });

  test('resolveDID', () async {
    final key = DIDKit.generateEd25519Key();
    final did = DIDKit.keyToDID('key', key);
    final resolutionResult = jsonDecode(DIDKit.resolveDID(did, '{}'));
    expect(resolutionResult['didDocument'], isNotEmpty);
  });

  test('dereferenceDIDURL', () async {
    final key = DIDKit.generateEd25519Key();
    final verificationMethod = DIDKit.keyToVerificationMethod('key', key);
    final derefResult = jsonDecode(DIDKit.dereferenceDIDURL(verificationMethod, '{}'));
    expect(derefResult, isList);
  });

  test('DIDAuth', () async {
    final key = DIDKit.generateEd25519Key();
    final did = DIDKit.keyToDID('key', key);
    final verificationMethod = DIDKit.keyToVerificationMethod('key', key);

    final challenge = Uuid().v4();
    final proofOptions = jsonEncode({
      'proofPurpose': 'assertionMethod',
      'verificationMethod': verificationMethod,
      'challenge': challenge
    });
    final vp = DIDKit.DIDAuth(did, proofOptions, key);
    final verifyOptions = jsonEncode({
      'proofPurpose': 'assertionMethod',
      'challenge': challenge
    });
    final verification = DIDKit.verifyPresentation(vp, verifyOptions);
    final verifyResult = jsonDecode(verification);
    expect(verifyResult['errors'], isEmpty);
  });
}

File: lib/java/test/com/spruceid/DIDKitTest.java
package com.spruceid;

import com.spruceid.DIDKit;

class DIDKitTest {
    public static void main(String[] args) throws DIDKitException {
        String version = DIDKit.getVersion();

        // Generate key
        String jwk = DIDKit.generateEd25519Key();

        // Convert key to DID
        String did = DIDKit.keyToDID("key", jwk);

        // Get verificationMethod for DID
        String verificationMethod = DIDKit.keyToVerificationMethod("key", jwk);

        // Trigger Exception
        boolean threw = false;
        try {
            DIDKit.keyToDID("key", "{}");
        } catch (DIDKitException e) {
            threw = true;
        }
        assert threw;

        // Issue Credential (LDP)
        String credential = "{"
            + "   \"@context\": \"https://www.w3.org/2018/credentials/v1\","
            + "   \"id\": \"http://example.org/credentials/3731\","
            + "   \"type\": [\"VerifiableCredential\"],"
            + "   \"issuer\": \"" + did + "\","
            + "   \"issuanceDate\": \"2020-08-19T21:41:50Z\","
            + "   \"credentialSubject\": {"
            + "       \"id\": \"did:example:d23dd687a7dc6787646f2eb98d0\""
            + "   }"
            + "}";
        String vcOptions = "{"
            + "  \"proofPurpose\": \"assertionMethod\","
            + "  \"verificationMethod\": \"" + verificationMethod + "\""
            + "}";
        String vc = DIDKit.issueCredential(credential, vcOptions, jwk);
        // Issue Credential (JWT)
        String vcOptionsJwt = "{"
            + "  \"proofPurpose\": \"assertionMethod\","
            + "  \"proofFormat\": \"jwt\","
            + "  \"verificationMethod\": \"" + verificationMethod + "\""
            + "}";
        String vcJwt = DIDKit.issueCredential(credential, vcOptionsJwt, jwk);

        // Verify Credential (LDP)
        String vcVerifyOptions = "{"
            + "  \"proofPurpose\": \"assertionMethod\""
            + "}";
        String vcResult = DIDKit.verifyCredential(vc, vcVerifyOptions);
        assert vcResult.contains("\"errors\":[]");

        // Verify Credential (JWT)
        vcVerifyOptions = "{"
            + "  \"proofPurpose\": \"assertionMethod\","
            + "  \"proofFormat\": \"jwt\""
            + "}";
        vcResult = DIDKit.verifyCredential(vcJwt, vcVerifyOptions);
        assert vcResult.contains("\"errors\":[]");

        // Issue Presentation (LDP)
        String presentation = "{"
            + "   \"@context\": [\"https://www.w3.org/2018/credentials/v1\"],"
            + "   \"id\": \"http://example.org/presentations/3731\","
            + "   \"type\": [\"VerifiablePresentation\"],"
            + "   \"holder\": \"" + did + "\","
            + "   \"verifiableCredential\": " + vc
            + "}";
        String vpOptions = "{"
            + "  \"proofPurpose\": \"authentication\","
            + "  \"verificationMethod\": \"" + verificationMethod + "\""
            + "}";
        String vp = DIDKit.issuePresentation(presentation, vpOptions, jwk);

        // Issue Presentation (JWT)
        String vpOptionsJwt = "{"
            + "  \"proofPurpose\": \"authentication\","
            + "  \"proofFormat\": \"jwt\","
            + "  \"verificationMethod\": \"" + verificationMethod + "\""
            + "}";
        String vpJwt = DIDKit.issuePresentation(presentation, vpOptions, jwk);

        // Verify Presentation (LDP)
        String vpVerifyOptions = "{"
            + "  \"proofPurpose\": \"authentication\""
            + "}";
        String vpResult = DIDKit.verifyPresentation(vp, vpVerifyOptions);
        assert vpResult.contains("\"errors\":[]");

        // Verify Presentation (JWT)
        vpVerifyOptions = "{"
            + "  \"proofPurpose\": \"authentication\","
            + "  \"proofFormat\": \"jwt\""
            + "}";
        vpResult = DIDKit.verifyPresentation(vpJwt, vpVerifyOptions);
        assert vpResult.contains("\"errors\":[]");

        // Resolve DID
        String resolutionResult = DIDKit.resolveDID(did, "{}");
        assert resolutionResult.contains("\"didDocument\":{");

        // Dereference DID URL
        String dereferencingResult = DIDKit.dereferenceDIDURL(verificationMethod, "{}");
        assert dereferencingResult.startsWith("[{");

        // Create a DIDAuth VP (LDP)
        vpOptions = "{"
            + "  \"proofPurpose\": \"authentication\","
            + "  \"domain\": \"example.org\","
            + "  \"verificationMethod\": \"" + verificationMethod + "\""
            + "}";
        vp = DIDKit.DIDAuth(did, vpOptions, jwk);

        // Create a DIDAuth VP (JWT)
        vpOptions = "{"
            + "  \"proofPurpose\": \"authentication\","
            + "  \"proofFormat\": \"jwt\","
            + "  \"domain\": \"example.org\","
            + "  \"verificationMethod\": \"" + verificationMethod + "\""
            + "}";
        vpJwt = DIDKit.DIDAuth(did, vpOptions, jwk);

        // Verify DIDAuth VP (LDP)
        vpVerifyOptions = "{"
            + "  \"domain\": \"example.org\","
            + "  \"proofPurpose\": \"authentication\""
            + "}";
        vpResult = DIDKit.verifyPresentation(vp, vpVerifyOptions);
        assert vpResult.contains("\"errors\":[]");

        // Verify DIDAuth VP (JWT)
        vpVerifyOptions = "{"
            + "  \"domain\": \"example.org\","
            + "  \"proofPurpose\": \"authentication\","
            + "  \"proofFormat\": \"jwt\""
            + "}";
        vpResult = DIDKit.verifyPresentation(vpJwt, vpVerifyOptions);
        assert vpResult.contains("\"errors\":[]");
    }
}
